File: server/toss-server.js
const net = require('net');
const { useHandlers, wAmount, useWriteQueue, sendChunks }  = require('../util/misc');
const { SIGNALS, SOCKET_EVENTS, LOG_STATES, LOG_TYPES, LOG_NAMES } = require('../util/constants');
const TossLogger = require('./toss-logger');
const Pillow = require('../pillow/index');
const Slip = require('../slip/index');


class TossServer extends net.Server {
  clients = [];
  _handlers = {};
  _handlersDir = null;
  static _defaultCId = 'tsId';
  _clientIndicator = '';

  constructor(
    { handlers = {}, handlersDir, clientIndicator } = {},
    options, connectionListener
  ) {
    super(
      options,
      client => {
        this.registerClient(client);
        if (connectionListener) {
          connectionListener(client);
        }
      }
    );

    this._handlers = handlers;
    this._handlersDir = handlersDir;
    this._clientIndicator = clientIndicator || TossServer._defaultCId;

    process.on(SIGNALS.SIGINT, () => this.close());
  }

  _getHandling() {
    const handling = this.clients.length;
    return `Handling ${wAmount(handling, 'connection')} now`;
  }

  registerClient(client) {
    const useTs = this._clientIndicator === TossServer._defaultCId;
    if (useTs) {
      client[TossServer._defaultCId] = (new Date()).getTime();
    }

    client.res = (...args) => this.res(client, ...args);
    client.err = (...args) => this.err(client, ...args);
    useWriteQueue(
      client,
      (state, queueSize, err) => {
        const waiting = `${queueSize ? wAmount(queueSize, 'item') : 'No items'} waiting to be sent`;
        const text = state === err
          ? `Last write resulted in an error: (${err.constructor.name}) ${err.message}`
          : waiting;
        let status;
        if (err) {
          status = TossLogger.status.error;
        } else {
          switch (state) {
            case LOG_STATES.written:
              status = TossLogger.status.success;
              break;
            case LOG_STATES.enqueued:
              status = TossLogger.status.prefix;
              break;
            case LOG_STATES.stopped:
              status = TossLogger.status.warn;
          }
        }
        TossLogger.log({ status, state, comment: text });
      }
    );

    this.clients.push(client);
    const connectionDescription = useTs
      ? `, ${TossServer._defaultCId} = ${client[TossServer._defaultCId]}` :
      '';
    TossLogger.log({
      status: TossLogger.status.success,
      comment: `Established a new connection${connectionDescription}. ${this._getHandling()}`
    });
    useHandlers(
      client, {
        makeExtraArgs: ev => [{ client, server: this, ev }],
        handlersDir: this._handlersDir,
        handlers: this._handlers,
        handledEvents: SOCKET_EVENTS,
        log: (ev, handler) => TossLogger.log({
          type: LOG_TYPES.Event,
          name: ev,
          state: handler ? LOG_STATES.passedToHandle : LOG_STATES.skipped,
          status: handler ? TossLogger.status.prefix : TossLogger.status.warn
        }),
        // Server errors (non-classified, happen because of bad code) are handled here
        // If there's an error with the server itself, not some single connected client,
        // all the connected clients will receive ECONNRESET
        catcherFunc: (err, ev) => {
          TossLogger.log({
            type: LOG_TYPES.Event,
            name: ev,
            state: LOG_STATES.error,
            status: TossLogger.status.error,
            comment: err.message
          });
          client.write(Slip.serialize({
            status: Pillow.responseStatus.ERR_SERVER.code,
            data: { errors: { _err: ['Server error'] } }
          }));
        }
      }
    );
  }

  unregisterClient(client) {
    this.clients = this.clients.filter(c => c !== client);
    TossLogger.log({
      status: TossLogger.status.info,
      comment: `Closed connection for ${client[this._clientIndicator]}. ${this._getHandling()}`
    });
  }

  res(client, action, status, data, files, cb) {
    const toSend = { action, status };
    if (data) {
      toSend.data = { ...data, time: new Date() };
    }
    const serializedData = Slip.serialize(toSend, files);
    sendChunks(
      client,
      serializedData,
      chunks => Slip.serialize({
        action: Pillow.actions.chunks,
        data: { chunks },
        status: Pillow.responseStatus.OK.code
      }),
      false, {
        cb,
        log: chunks => TossLogger.log({
          type: LOG_TYPES.Chunks,
          name: LOG_NAMES.chunksSent,
          comment: `Enqueued ${chunks} bytes of data to be sent`
        })
      }
    );
  }

  err(client, action, errors, status, cb) {
    const serializedData = Slip.serialize({ action, data: { errors }, status });
    client.write(serializedData, cb);
  }

  broadcast(
    action, {
      filterClients = list => list,
      getData = () => null,
      files = {},
      getStatus = () => Pillow.responseStatus.OK.code
    } = {}
  ) {
    const writeableClients = this.clients.filter(c =>
      ['open', 'writeOnly'].includes(c.readyState) &&
      !!c[this._clientIndicator]
    );
    const filteredClients = filterClients(writeableClients);
    let amount = filteredClients.length;
    let errorCount = 0;

    TossLogger.log({
      comment: `Broadcasting to ${wAmount(amount, 'client')}...`,
      status: TossLogger.status.prefix
    });

    const checkFinish = (c, err) => {
      if (err) {
        errorCount++;
        // Don't exactly know who to acknowledge when a message did not broadcast properly,
        // so let's keep this server-side-only info
        TossLogger.log({
          status: TossLogger.status.error,
          comment: `Error sending data to ${c[this._clientIndicator]}`
        });
      }
      if (--amount === 0) {
        TossLogger.log({
          comment: `Finished broadcasting with ${wAmount(errorCount, 'error')}`,
          status: errorCount ? TossLogger.status.warn : TossLogger.status.success
        });
      }
    }

    const time = new Date();
    filteredClients.forEach(c => {
      const toSend = { action, status: getStatus(c) };
      const data = getData(c);
      if (data) {
        toSend.data = { ...data, time };
      }
      const serializedData = Slip.serialize(toSend, { data: files });
      sendChunks(
        c,
        serializedData,
        chunks => Slip.serialize({
          action: Pillow.actions.chunks,
          data: { chunks },
          status: Pillow.responseStatus.OK.code
        }),
        false, {
          cb: err => checkFinish(c, err),
          log: chunks => TossLogger.log({
            type: LOG_TYPES.Chunks, name: LOG_NAMES.chunksSent,
            comment: `Enqueued ${chunks} bytes of data to be sent to ${c[this._clientIndicator]}`
          })
        }
      );
    });
  }

  listen(port, hostname, backlog, listeningListener) {
    return super.listen(port, hostname, backlog, (...args) => {
      const { address, port } = this.address();
      TossLogger.log({ comment: `Started the server on ${address}:${port}`, status: TossLogger.status.info });
      if (listeningListener) {
        listeningListener(...args);
      }
    });
  }

  close(callback) {
    const waitingFor = this.clients.length;
    const { address, port } = this.address();
    if (waitingFor) {
      TossLogger.log({
        comment: `Closing the server, waiting for ${wAmount(waitingFor, 'client')} to disconnect...`,
        status: TossLogger.status.prefix
      });
    }
    return super.close(() => {
      TossLogger.log({
        comment: `Closed the server on ${address}:${port}`,
        status: TossLogger.status.info
      });
    });
  }
}

module.exports = TossServer;

File: server/toss_server.py
import socket
from signal import SIGINT, SIGTERM, signal
import asyncio
import logger
from toss_client_holder import TossClientHolder
from typing import List, Callable
from pillow import Actions, ResponseStatus
import util.misc as _


class GracefulExit(SystemExit):
    code = 1


class TossServer:
    clients: List[TossClientHolder] = []

    def __init__(self, address, port):
        self.address = address
        self.port = port

    def get_running_on(self) -> str:
        return ', '.join(f'{sock.getsockname()[0]}:{sock.getsockname()[1]}' for sock in self.sock.sockets)

    def __raise_graceful_exit(self, *_):
        self.sock.get_loop().stop()
        raise GracefulExit()

    def get_handling_now(self):
        return f'Handling {_.w_amount(len(self.clients), "connection")} now'

    async def broadcast(
            self, action: str,
            filter_clients: Callable[[List[TossClientHolder]], List[TossClientHolder]] = lambda lst: lst,
            get_data: Callable[[TossClientHolder], dict] = lambda *_: None,
            files: dict = None,
            get_status: Callable[[TossClientHolder], int] = lambda *_: ResponseStatus.OK.value,
            cb: Callable = lambda: None
    ):
        if files is None:
            files = {}

        writeable_clients = list(filter(lambda c: not c.writer.is_closing(), self.clients))
        filtered_clients = filter_clients(writeable_clients)
        amount = len(filtered_clients)
        error_count = 0

        logger.log(comment=f'Broadcasting to {_.w_amount(amount, "client")}...', status=logger.Status.prefix)

        time = _.now()  # So that the time is the same for each client
        for client in filtered_clients:
            to_send = {'action': action, 'status': get_status(client)}
            data = get_data(client)
            if data is not None:
                to_send['data'] = data
                to_send['data']['time'] = time
            try:
                await client.write_safely(to_send, files)
            except IOError as err:  # For the rare case when some client closes the connection while broadcasting
                logger.log(
                    status=logger.Status.error,
                    comment=f'Error sending data to {client}: {str(err)}',
                )
                error_count += 1

        logger.log(
            comment=f'Finished broadcasting with {_.w_amount(error_count, "error")}',
            status=logger.Status.warn if error_count else logger.Status.success
        )
        cb()

    async def close_completely(self, err: Exception = None):
        if err:
            logger.log(
                occasion_type=logger.OccasionType.error.value,
                occasion_name=_.get_clean_type(err),
                status=logger.Status.error,
                comment=str(err)
            )

        waiting_for = len(self.clients)
        if waiting_for:
            logger.log(
                comment=f'Closing the server, waiting for {_.w_amount(waiting_for, "client")} to disconnect...',
                status=logger.Status.prefix
            )
            await self.broadcast(
                action=Actions.close_server.value,
                get_status=lambda *_: ResponseStatus.OK_EMPTY.value
            )

            for client in self.clients:
                if not client.writer.is_closing():
                    client.writer.close()
                    await client.writer.wait_closed()

        self.sock.close()
        await self.sock.wait_closed()
        logger.log(comment=f'Toss Server has been closed', status=logger.Status.info)

    async def __register_client(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        client = TossClientHolder(reader, writer, self, len(self.clients) + 1)
        self.clients.append(client)
        logger.log(
            status=logger.Status.success,
            comment=f'Established a new connection. {self.get_handling_now()}'
        )
        try:
            await client.run()
        except BaseException as err:  # All the known exceptions are handled within the Client
            logger.log(
                occasion_type=logger.OccasionType.error.value,
                occasion_name=_.get_clean_type(err),
                status=logger.Status.error,
                comment=f'{client}: {str(err)}'
            )
            if not writer.is_closing():
                to_send = {
                    'status': ResponseStatus.ERR_SERVER.value,
                    'data': {'errors': {'_err': 'Server error'}}
                }
                await client.write_safely(to_send)

    async def unregister_client(self, client):
        self.clients = _.l_filter(lambda c: c != client, self.clients)
        logger.log(
            status=logger.Status.info,
            comment=f'Closed connection for {client}. {self.get_handling_now()}'
        )
        await self.broadcast(action=Actions.log_out.value, get_data=lambda *_: {'username': client.username})

    async def __run(self):
        # AF_INET means we use address + port for address
        # SOCK_STREAM means this is a TCP socket - set by default internally
        self.sock = await asyncio.start_server(
            client_connected_cb=self.__register_client,
            host=self.address,
            port=self.port,
            family=socket.AF_INET,
            start_serving=False
        )

        signal(SIGINT, self.__raise_graceful_exit)
        signal(SIGTERM, self.__raise_graceful_exit)
        running_on = self.get_running_on()
        logger.log(comment=f'Toss Server is running on {running_on}', status=logger.Status.info)

        try:
            await self.sock.serve_forever()
        except GracefulExit:
            pass
        finally:
            await self.close_completely()

    def run(self):
        asyncio.run(self.__run())
